### PR TITLE
[native assets] Rewrite install names for relocated native libraries

### PR DESCRIPTION
**Pull request by blaugold**
Native libraries that are contributed by native asset builders can depend on each other. For macOS and iOS, native libraries are repackaged into Frameworks, which renders install names that have been written into dependent libraries invalid. 

With this change, a mapping between old and new install names is maintained, and install names in dependent libraries are rewritten as a final step.

Related to https://github.com/dart-lang/native/issues/190

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:225]: These tests LGTM!

I think we should also exercise this behavior in an integration test:

* test/integration.shard/isolated/native_assets_test.dart

You could write a function near `addLinkHookDepedendency` that modifies the existing C code to depend on another dylib. And change the `hook/build.dart` to produce two dylibs.

This will also ensure we're testing it on the other OSes where no changes are needed.

**Review by dcharkes — COMMENTED** [review]
lgtm on the overall approach. Thanks @blaugold! 🎸 

It seems like the integration test is failing. I believe we should exercise this behavior in the integration test as well. You can run it with:

`cd <...>/flutter/packages/flutter_tools/ && flutter test test/integration.shard/isolated/native_assets_test.dart` (takes quite a while to run)

File: packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart
Comment (blaugold) [reply packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:225]: Testing linking in the integration tests on Windows is blocked by support for passing linker flags to MSVC through `CBuilder`.

Comment (dcharkes) [reply packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:225]: 🙏 

We (@mosuem me) haven't found the time to look at Windows yet. If you want, you can embrace the yak shave and make a PR on that repo.

I'm also fine with filing an issue on that repo, and linking it here on the reason why it's skipped.

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:726]: Nit: It would be slightly cleaner if the return value would come from the linked method. (Though, this will also definitely let the runtime fail with dynamic linking error if it can't find the dylib.)

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:733]: If we're calling the linked dylib from C code, we don't need a Dart bindings for it right?

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:752]: Same feedback as on https://github.com/dart-lang/native/pull/1413: `final builders = [`

File: packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:309]: indentation?

File: packages/flutter_tools/test/general.shard/isolated/macos/native_assets_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/general.shard/isolated/macos/native_assets_test.dart:362]: `else` (and maybe invert order to get rid of `!`)

Or rewrite completely to `if (flutterTester) ... [` ... `] else ... [` ... `]`.

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:89]: Update documentation to include also rewriting names of dependencies.

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:168]: Could be moved into a separate function and have unit tests in `test/general.shard/isolated/macos/native_assets_host_test.dart`. (In case we run into future cases of `otool` that have slightly different output and want to ensure we keep parsing all versions.)

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:208]: Is this a param that can be provided multiple times? If so, we should rewrite the function so that it takes a map or a list of tuples. Then we have less process invocations.

**Review by dcharkes — COMMENTED** [review]
❤️

**Review by blaugold — COMMENTED** [review]
It turns out that linking to a dynamic library on Windows is more straightforward than I thought. It's enough to pass the `lib` associated with the `dll` to the compiler, and no additional linker flags are necessary. I've updated the integration test to also run on Windows. With the Flutter embedder, loading the linked library works, both for a built desktop app and the flutter tester. 

Using the same builder hook setup as with Flutter but with standalone Dart, the linked library can never be found. I've tested `dart run`, `dart test` and `dart build`, always with the same result. I'm going to work on https://github.com/dart-lang/native/issues/1419, but I would prefer landing this PR first and coming back later to update the tests.

@dcharkes If you have any idea what differences between embedders could cause loading the linked library to fail with standalone Dart, that would be very helpful for https://github.com/dart-lang/native/issues/1419.

**Comment by dcharkes** [comment]
> It turns out that linking to a dynamic library on Windows is more straightforward than I thought. It's enough to pass the `lib` associated with the `dll` to the compiler, and no additional linker flags are necessary. I've updated the integration test to also run on Windows. With the Flutter embedder, loading the linked library works, both for a built desktop app and the flutter tester.

👍 

> Using the same builder hook setup as with Flutter but with standalone Dart, the linked library can never be found. I've tested `dart run`, `dart test` and `dart build`, always with the same result. I'm going to work on [dart-lang/native#1419](https://github.com/dart-lang/native/issues/1419), but I would prefer landing this PR first and coming back later to update the tests.
> 
> @dcharkes If you have any idea what differences between embedders could cause loading the linked library to fail with standalone Dart, that would be very helpful for [dart-lang/native#1419](https://github.com/dart-lang/native/issues/1419).

I've replied in https://github.com/dart-lang/sdk/issues/56459.

**Review by dcharkes — APPROVED** [review]
LGTM from my side, but I'd like approval from some Flutter folks as well before merging.

**Comment by dcharkes** [comment]
@vashworth These are mostly iOS/MacOS changes, could you please take a look at this PR as well.

And @christopherfujino for a sign off on this flutter_tools PR.

**Comment by blaugold** [comment]
@dcharkes I realized that I broke the integration test when refactoring to return the result of the linked library call. The call to the linked library was only made in release mode. The flutter-tester device is only exercised in debug mode, so there was no call to the linked library. Now the test calls the linked library in both modes and that revealed that flutter-tester has the same issue loading relative DLLs on Windows as standalone Dart.

**Comment by dcharkes** [comment]
> Now the test calls the linked library in both modes and that revealed that flutter-tester has the same issue loading relative DLLs on Windows as standalone Dart.

Yes that makes sense. We don't do any "bundling" for flutter-tester but try to just run in place.

**Comment by christopherfujino** [comment]
@andrewkolos can you review this?

**Comment by blaugold** [comment]
@dcharkes I've added rewriting install names for flutter tester. Please take another look. Thanks!

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:89]: Is this TODO still correct? You resolved the Windows linking down below.

**Review by dcharkes — APPROVED** [review]
All looks good!

(Context: https://github.com/dart-lang/native/issues/190#issuecomment-2288090460)

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart
Comment (blaugold) [reply packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:146]: One can dream 😅
Done

File: packages/flutter_tools/test/general.shard/isolated/build_system/targets/native_assets_test.dart
Comment (blaugold) [reply packages/flutter_tools/test/general.shard/isolated/build_system/targets/native_assets_test.dart:160]: Done

File: packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart
Comment (blaugold) [reply packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:255]: Done

File: packages/flutter_tools/test/general.shard/isolated/build_system/targets/native_assets_test.dart
Comment (blaugold) [reply packages/flutter_tools/test/general.shard/isolated/build_system/targets/native_assets_test.dart:127]: Done

**Review by andrewkolos — APPROVED** [review]
LGTM with nit once @dcharkes' items are addressed. I didn't get to do a deep dive here since I'm preoccupied with tool flakes and crashes, but I was able to understand the issue and the PR description. The approach looks good to me.

File: packages/flutter_tools/lib/src/tester/flutter_tester.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/tester/flutter_tester.dart:156]: Doesn't the build function do this already if the param is not passed?

File: packages/flutter_tools/test/general.shard/flutter_tester_device_test.dart
Comment (dcharkes) [inline packages/flutter_tools/test/general.shard/flutter_tester_device_test.dart:15]: This would break the roll into G3.

Instead add a new test in `test/general.shard/isolated/` that dir is excluded in the import.

For more info see: https://github.com/flutter/flutter/pull/143055

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/native_assets.dart:5]: We can't have imports to `package:native_assets_cli` outside the isolate directory. Neither can we import directly from the `isolated` directory.

Inside g3, Flutter doesn't use the "native assets" implementation with hooks.

For more documentation, see: https://github.com/flutter/flutter/pull/143055

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/native_assets.dart:25]: Move this into `isolated/`, possibly in `TestCompilerNativeAssetsBuilder`.

**Review by dcharkes — COMMENTED** [review]
The current PR adds imports to `native_assets_builder` and `native_assets_cli` in files that are not either in (1) `isolated/` or (2) a `main` function. That will break the roll of Flutter with other build systems. For more info see https://github.com/flutter/flutter/pull/143055.

Maybe you can add `windowsPathWithNativeAssetsBuildDirectory` to `TestCompilerNativeAssetsBuilder`. Otherwise, we'd need to add a new interface that we plumb through from the main functions.

(LGTM to all other changes.)

File: packages/flutter_tools/lib/src/globals.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/globals.dart:310]: I believe introducing new globals might be prohibited, I remember threading through the objects from main everywhere when introducing the indirection.

cc @christopherfujino

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (dcharkes) [inline packages/flutter_tools/lib/src/native_assets.dart:21]: Maybe it's slightly cleaner to simply return the path here and do the prepending on the existing env PATH at the place where you pass it to any environment.

**Review by dcharkes — APPROVED** [review]

File: packages/flutter_tools/lib/src/globals.dart
Comment (blaugold) [reply packages/flutter_tools/lib/src/globals.dart:310]: I did it this way because I need to pass `nativeAssetsBuilder` to `FlutterDeviceManager` in `context_runner.dart` and wanted to keep all instantiations of `TestCompilerNativeAssetsBuilderImpl` in `exectubale.dart` and imports to `isolated` out of `context_runner.dart`. But that might not be the best approach. 😅

File: packages/flutter_tools/lib/src/native_assets.dart
Comment (blaugold) [reply packages/flutter_tools/lib/src/native_assets.dart:21]: Is this class implemented/extended somewhere in g3 so that adding an abstract method would break it?

Comment (dcharkes) [reply packages/flutter_tools/lib/src/native_assets.dart:21]: Nope, adding methods should be fine. 👍

File: packages/flutter_tools/lib/src/globals.dart
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/globals.dart:310]: > I believe introducing new globals might be prohibited

This was the truth at one point, but now it's complicated. Context: https://github.com/flutter/flutter/issues/47161. TL;DR I no longer prohibit globals when they make sense, but I would like to have globals dependency injected through a single DI object (similar to `BuildContext` in flutter widgets) rather than using zones. This is outside the scope of this PR.

**Review by dcharkes — APPROVED** [review]

**Comment by dcharkes** [comment]
Thanks a lot @blaugold ! 🚀

**Pull request merged**
